src/array/cpu/spmm_blocking_libxsmm.h

```diff
@@ -0,0 +1,517 @@
+            const IdType row_end   = my_cur_col_id[(i - M_start) * 2 + 1];
+            cur_csr_indptr[i - M_start] = cur_nnz;
+            IdType eid;
+            for (eid = row_start; eid < row_end; eid++) {
+              const IdType src = Indices[eid];
+              const IdType edge = Edges[eid];
+              if (src >= K_end) {
+                break;
+              }
+              CHECK_LT(cur_indices_id + cur_nnz, nnz);
+              if (use_lhs)
+                cur_csr_indices[cur_nnz] = src;
+              if (use_rhs)
+                cur_csr_edges[cur_nnz] = edge;
+              cur_nnz++;
+            }
+            my_cur_col_id[(i - M_start) * 2] = eid;
+          }
+          cur_csr_indptr[cur_csr.num_rows] = cur_nnz;
+          cur_indices_id += cur_nnz;
+          cur_csr.indptr = cur_csr_indptr;
+          if (use_lhs)
+            cur_csr.indices = cur_csr_indices;
+          if (use_rhs)
+            cur_csr.data = cur_csr_edges;
+          block_csr_array[m * num_K_blocks + k] = cur_csr;
+
+        }
+        CHECK_EQ(nnz, cur_indices_id);
+      }
+      free(my_cur_col_id);
+    }
+  }
+  else {
+#pragma omp for
+    for (IdType m = 0; m < num_M_blocks; m++) {
+      IdType M_start = m * M_block_size;
+      IdType M_end = (m + 1) * M_block_size;
+      if (M_end > M) M_end = M;
+
+      CSRMatrixInternal<IdType, IdType> cur_csr;
+      cur_csr.num_rows = M_end - M_start;
+      cur_csr.num_cols = K;
+      cur_csr.indptr = IndPtr + M_start;
+      cur_csr.indices = Indices;
+      cur_csr.data = Edges;
+
+      block_csr_array[m] = cur_csr;
+    }
+  }
+}
+
+
+template <typename IdType, typename DType, typename Op>
+inline libxsmm_meltwfunction_opreduce_vecs_idx SpMMCreateLibxsmmKernel(
+    bool has_idx,
+    IdType N,
+    int32_t _ld,
+    libxsmm_meltw_opreduce_vecs_flags redop_flag,
+    bool is_cmp) {
+
+  libxsmm_meltw_opreduce_vecs_flags opredop_flags;
+  if (std::is_same<Op, op::Add<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_ADD;
+  }
+  else if (std::is_same<Op, op::Sub<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_SUB;
+  }
+  else if (std::is_same<Op, op::Mul<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_MUL;
+  }
+  else if (std::is_same<Op, op::Div<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_DIV;
+  }
+  else if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
+    opredop_flags = LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OP_COPY;
+  }
+  if (std::is_same<Op, op::CopyLhs<DType>>::value) {
+    opredop_flags = (libxsmm_meltw_opreduce_vecs_flags)(opredop_flags |
+                     LIBXSMM_MELTW_FLAG_OPREDUCE_VECS_OPORDER_VECIDX_VECIN);
+  }
+  else if (std::is_same<Op, op::CopyRhs<DType>>::value) {
```

Comment:
Is there any reason to duplicate the if statement from line 191?

Reply:
Yes, I have added comments for the same in the code.